### PR TITLE
[ONNX] De-duplicate initializers

### PR DESCRIPTION
**Pull request by BowenBao**
ScriptModule export introduces duplicated ONNX initializers for shared weights, unnecessarily increases ONNX model size. This PR de-duplicates ONNX initializers for model exported in eval mode, by checking if the underlying tensors share the same `data_ptr`, `strides` and `sizes`.

### CODE REVIEW

File: torch/csrc/jit/passes/onnx/helper.h [hunk 1]
```diff
@@ -14,6 +14,8 @@ static const int OPSET_VERSION_10 = 10;
 static const int OPSET_VERSION_11 = 11;
 static const int OPSET_VERSION_12 = 12;
 static const int OPSET_VERSION_13 = 13;
+static const int OPSET_VERSION_14 = 14;
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/helper.h:17]: Not needed now, but maybe we should remove them at some point. Maybe use the constants defined in onnx.proto instead if we don't want to just use int literals.

**Review by garymm — CHANGES_REQUESTED** [review]

File: test/onnx/test_utility_funs.py [hunk 2]
```diff
@@ -1127,6 +1127,34 @@ def forward(self, x):
...
+                self.layer3 = self.layer1
+
+                # Reusing parameters.
+                self.layer2.weight = self.layer1.weight
+                self.layer1.bias = self.layer2.bias
+
+            def forward(self, x):
+                return self.layer3(self.layer2(self.layer1(x)))
+
+        model = torch.jit.script(MyModule())
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply test/onnx/test_utility_funs.py]: The test case was named with `torchscript` in the postfix, for testing script module case. (Updated PR covers both nn.Module and ScriptModule). Is this sufficient, or are you asking to explain in the comment why scripting needs to be tested?

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp [hunk 3]
```diff
@@ -142,9 +142,63 @@ static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
   }
 }
 
+static void DeduplicateParameters(
+    Block* b,
+    ValueToParamPairMap& valsToParamsMap) {
+  // Parameters are considered as duplicated if their data_ptr, strides and
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/eval_peephole.cpp]: Left there as there were considerations to use different logic to determinate duplication, but refactored away now.

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp [hunk 4]
```diff
@@ -142,9 +142,63 @@ static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
...
+    auto it = std::find_if(
+        uniqueVals.begin(), uniqueVals.end(), is_same_tensor_as(v));
+    if (it == uniqueVals.end()) {
+      uniqueVals.emplace_back(v);
+    } else {
+      inputsIndicesToRemove.emplace_back(i);
+      v->replaceAllUsesWith(*it);
+    }
+  }
+  for (auto it = inputsIndicesToRemove.rbegin();
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/eval_peephole.cpp]: Yes, but it won't make the code cleaner. `graph.eraseInput(size_t i)` only takes index as argument, so going backwards make tracking indices difficult (duplication might be found earlier for later discovered initializer, altering indices of previously discovered initializers).

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp [hunk 5]
```diff
@@ -142,9 +142,63 @@ static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
...
+    g->eraseInput(*it);
+  }
+}
+
 void EvalPeepholeONNX(Block* b, ParamMap& paramsDict) {
+  auto opset_version = GetOpsetVersion();
   auto valsToParamsMap = buildValueToParamsMap(b, paramsDict);
   fuseConvBatchNorm(b, valsToParamsMap);
+  if (opset_version >= OPSET_VERSION_15) {
+    // Apply de-duplication after opset 15 for backward compatibility.
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/eval_peephole.cpp]: This is open to discussion. Not related to ONNX release, but we are changing default behavior of exporter regarding initializers, so I'm wondering if we should only apply it on newer opsets for BC purpose. This applies for other future passes of similar types as well. Would love to hear your thoughts on this.

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp [hunk 6]
```diff
@@ -142,9 +142,63 @@ static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
...
+       ++it) {
+    g->eraseInput(*it);
+  }
+}
+
 void EvalPeepholeONNX(Block* b, ParamMap& paramsDict) {
+  auto opset_version = GetOpsetVersion();
   auto valsToParamsMap = buildValueToParamsMap(b, paramsDict);
   fuseConvBatchNorm(b, valsToParamsMap);
+  if (opset_version >= OPSET_VERSION_15) {
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/eval_peephole.cpp]: updated to argument.

File: test/onnx/test_utility_funs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (garymm) [reply test/onnx/test_utility_funs.py]: LGTM.

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (garymm) [reply torch/csrc/jit/passes/onnx/eval_peephole.cpp]: I think unless we have reason to believe this will not work on older runtimes, we should not guard the behavior.
I suggest we test this against Caffe2 using test_pytorch_onnx_caffe2.py. If we want to be extra careful you could also install an older version of ORT and run test_pytorch_onnx_onnxruntime.py.
If it passes both of those, then I think it's likely that we don't need to guard this behavior.

WDYT?

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (garymm) [reply torch/csrc/jit/passes/onnx/eval_peephole.cpp]: OK.

File: torch/csrc/jit/passes/onnx/deduplicate_initializers.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+namespace jit {
+
+namespace onnx {
+using namespace ::c10::onnx;
+}
+
+void DeduplicateInitializers(
+    std::shared_ptr<Graph>& g,
+    ValueToParamPairMap& valsToParamsMap,
+    int opset_version,
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/deduplicate_initializers.cpp:17]: Seems this is not used now.

File: torch/csrc/jit/passes/onnx/deduplicate_initializers.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    std::map<std::string, IValue>& paramsDict,
+    bool is_train,
+    int opset_version) {
+  auto valsToParamsMap = buildValueToParamsMap(g->block(), paramsDict);
+  // ONNX spec does not support parameters with shared memory.
+  // This pass de-duplicate those parameters. Training is not affected.
+  DeduplicateInitializers(
+      g, valsToParamsMap, opset_version, DeduplicateInitializersByDataPtr);
+  if (!is_train) {
+    // More aggressive parameters de-duplication based on tensor values.
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/deduplicate_initializers.cpp:82]: Can you please rewrite the comment to explain why `ByValue` is not safe for training?
Something like:
```When training, a runtime may <fill me in>```

File: torch/csrc/jit/python/init.cpp [hunk 9]
```diff
@@ -215,6 +216,20 @@ void initJITBindings(PyObject* module) {
...
+          [](std::shared_ptr<Graph>& graph,
+             std::map<std::string, IValue>& paramsDict,
+             bool is_train,
+             int opset_version) {
+            DeduplicateInitializers(
+                graph,
+                paramsDict,
+                is_train,
+                opset_version); // overload resolution
+            return paramsDict;
```
Comment (garymm) [inline torch/csrc/jit/python/init.cpp:230]: A bit confusing to both modify the input and return something: seeing a return value suggests this is a pure function.
I suggest we remove the return value so it's clearer that the input is modified.
Or don't modify the input and make a copy if you prefer.

**Review by garymm — APPROVED** [review]
Please address comments before merging, but basically LGTM.
Thanks!

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/eval_peephole.cpp]: Ok we can give a try. For general usage I'm not too worried. If there are issues they are considered bugs and we can resolve them. I'm slightly worried against user custom behaviors on top of initializer/parameters.

File: torch/csrc/jit/python/init.cpp (hunk 9 above)
Comment (BowenBao) [reply torch/csrc/jit/python/init.cpp:230]: Updated, I thought this was pybind restriction, from comments of other places in the file. But it seems just passing reference works.

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/eval_peephole.cpp]: The `test_operator.py` result is one example, the change in parameters may appear confusing for users. For better readability I updated the logic to insert identity node instead of directly replacing initializers.

**Comment by garymm** [comment]
Still LGTM with identity nodes. Feel free to merge once tests pass.

File: torch/csrc/jit/passes/onnx/deduplicate_initializers.cpp [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+void DeduplicateInitializers(
+    std::shared_ptr<Graph>& g,
+    ValueToParamPairMap& valsToParamsMap,
+    bool (*comp)(at::Tensor&, at::Tensor&)) {
+  auto is_same_tensor_as = [&valsToParamsMap, comp](Value* v1) {
+    return [&valsToParamsMap, v1, comp](Value* v2) {
+      if ((valsToParamsMap.find(v1) == valsToParamsMap.end()) ||
+          (valsToParamsMap.find(v2) == valsToParamsMap.end())) {
+        return false;
```
Comment (fxmarty) [inline torch/csrc/jit/passes/onnx/deduplicate_initializers.cpp:22]: @BowenBao Hi, I was wondering why the last initializer is treated differently and never deduplicated? cc @xenova

Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/deduplicate_initializers.cpp:22]: Hi @fxmarty, the code was sanity check to ensure the values are initializers (can be found in `valsToParamsMap`). Excluding last initializer is not the goal, did you find a model that exhibits such behavior? Is it related to #108342?

Comment (fxmarty) [reply torch/csrc/jit/passes/onnx/deduplicate_initializers.cpp:22]: Yes related to https://github.com/pytorch/pytorch/issues/108342. I think I misread the code, not super familiar with C++.
